crates/consensus/src/engine/mod.rs

```diff
@@ -291,3 +321,148 @@ where
+                result_rx.await,
+                Ok(Err(EngineApiError::TerminalTD { execution, consensus }))
+                    if execution == config.merge_terminal_total_difficulty.into()
+                        && consensus == transition_config.terminal_total_difficulty.into()
+            );
+        }
+
+        #[tokio::test]
+        async fn terminal_block_hash_mismatch() {
+            let (tx, rx) = unbounded_channel();
+            let client = Arc::new(MockEthProvider::default());
+            let config = Config::default();
+            let engine = EthConsensusEngine {
+                client: client.clone(),
+                config: config.clone(),
+                local_store: Default::default(),
+                rx: UnboundedReceiverStream::new(rx),
+            };
+
+            tokio::spawn(engine);
+
+            let terminal_block_number = 1000;
+            let consensus_terminal_block = random_block(terminal_block_number, None, None);
+            let execution_terminal_block = random_block(terminal_block_number, None, None);
+
+            let transition_config = TransitionConfiguration {
+                terminal_total_difficulty: config.merge_terminal_total_difficulty.into(),
+                terminal_block_hash: consensus_terminal_block.hash(),
+                terminal_block_number,
+            };
+
+            // Unknown block number
+            let (result_tx, result_rx) = oneshot::channel();
+            tx.send(EngineMessage::ExchangeTransitionConfiguration(
+                transition_config.clone(),
+                result_tx,
+            ))
+            .expect("failed to send engine msg");
+
+            assert_matches!(
+                result_rx.await,
+                Ok(Err(EngineApiError::TerminalBlockHash { execution, consensus }))
+                    if execution.is_none()
+                        && consensus == transition_config.terminal_block_hash
+            );
+
+            // Add block and to provider local store and test for mismatch
+            client.add_block(
+                execution_terminal_block.hash(),
+                execution_terminal_block.clone().unseal(),
+            );
+
+            let (result_tx, result_rx) = oneshot::channel();
+            tx.send(EngineMessage::ExchangeTransitionConfiguration(
+                transition_config.clone(),
+                result_tx,
+            ))
+            .expect("failed to send engine msg");
+
+            assert_matches!(
+                result_rx.await,
+                Ok(Err(EngineApiError::TerminalBlockHash { execution, consensus }))
+                    if execution == Some(execution_terminal_block.hash())
+                        && consensus == transition_config.terminal_block_hash
+            );
+        }
+
+        #[tokio::test]
+        async fn configurations_match() {
+            let (tx, rx) = unbounded_channel();
+            let client = Arc::new(MockEthProvider::default());
```

Comment:
Add helper for instantiating this and the channels etc given its the same in all tests